Benja press mention rework 901 905

Shriukan33 opened the pull request:
this PR : 

* Adds Unique constraint to PressMention's URLs, so no URL can be saved twice.
* Drops copyright warning that was too important on screen while irrelevant
* Add a pre-creation form that will only require the Article's URL you want to submit, and fetch OG data in it to pre-fill the creation form.
* Creation form's content now trumps OG data, so what the user inputs is showed, while OG is now suggesting content automatically.

Closes #901 #902 #905

JeffAbrahamson commented:
Why is there a hard minimum on this field?

![image](https://user-images.githubusercontent.com/1220683/187193375-f1cad359-f0d1-49a0-b5a0-d929771bfb45.png)

JeffAbrahamson commented:
This is really confusing.  I changed the URL.  If I can do that, that means you're ready to change your processing.  And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.

If the user changes the URL, there needs to be feedback about what happened.  It has to be transparent for the user.  For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?

![image](https://user-images.githubusercontent.com/1220683/187193688-40d9607c-2a53-46b9-8e1c-43545cb655b5.png)

JeffAbrahamson commented:
You're not fetching or storing og:site_name.

JeffAbrahamson commented:
* You're not creating a new publication, you're creating a new citation.  Or something like that.
* The journal name is missing (og_site_name).
* The publication date is today's date.  If you don't have something like `og:article:published_time`, then you don't know the date and you shouldn't provide it for the user.

A reminder, open graph is a protocol, it has a [specification](https://opengraphprotocol.org/).

![image](https://user-images.githubusercontent.com/1220683/187194764-4f9d522b-bd67-4264-8960-2bfbc9212079.png)

JeffAbrahamson commented:
I have yet to find an article under 200 characters.

![image](https://user-images.githubusercontent.com/1220683/187195665-36fd9575-2785-4bf2-a3f9-47cad0fbfc11.png)

JeffAbrahamson commented:
This should open in a new tab.  We do not want to take people away from our site.

http://localhost:8000/presse/admin/detail/8

JeffAbrahamson commented:
* These should also open in new tabs.
* The journal title is wrong, it is the base of the URL (which may or may not be meaningful), whereas you stored `newspaper_name` in the db.
* The images should be clickable, that is _the_ hotspot for clicks and it is dead here.
* Only four articles at a time on a big screen?  Is that reasoned?
* Where are the article descriptions (possibly truncated to N words with ellipsis)?

![image](https://user-images.githubusercontent.com/1220683/187195964-ad8e4058-00d5-4dd9-8f68-e16ef03828bf.png)

Shriukan33 commented:
> This is really confusing. I changed the URL. If I can do that, that means you're ready to change your processing. And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.
> 
> If the user changes the URL, there needs to be feedback about what happened. It has to be transparent for the user. For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?
> 
> Why does it say "www.lemonde.fr", that's neither the name of the journal nor stored in the db?
> 
> ![image](https://user-images.githubusercontent.com/1220683/187193688-40d9607c-2a53-46b9-8e1c-43545cb655b5.png)

I'm confused about this one, I don't understand what you went through, what you did, what you expected, what happened ?

JeffAbrahamson commented:
> > This is really confusing. I changed the URL. If I can do that, that means you're ready to change your processing. And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.
> > If the user changes the URL, there needs to be feedback about what happened. It has to be transparent for the user. For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?
> > Why does it say "www.lemonde.fr", that's neither the name of the journal nor stored in the db?
> > ![image](https://user-images.githubusercontent.com/1220683/187193688-40d9607c-2a53-46b9-8e1c-43545cb655b5.png)
> 
> I'm confused about this one, I don't understand what you went through, what you did, what you expected, what happened ?

The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.

Shriukan33 commented:
> The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.

So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?

JeffAbrahamson commented:
> > The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.
> 
> So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?

Did you try it?

I didn't even save.  I entered the PO URL, it did the fetch.  I changed the URL to LM, then saved.

Shriukan33 commented:
> > > The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.
> > 
> > 
> > So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?
> 
> Did you try it?
> 
> I didn't even save. I entered the PO URL, it did the fetch. I changed the URL to LM, then saved.

the fetch happens when you enter the url in the pre-form, to pre-fill the fields. Any change past that point is user input and trumps whatever OG was saying.
So yeah, if you fetch OG image and then change the URL, it keeps the old one because the fetch happened earlier. Why would you do a fetch on a url, have its og data prefill, only to change everything after ? It doesn't make sense ?

JeffAbrahamson commented:
> > I didn't even save. I entered the PO URL, it did the fetch. I changed the URL to LM, then saved.
> 
> the fetch happens when you enter the url in the pre-form, to pre-fill the fields. Any change past that point is user input and trumps whatever OG was saying. So yeah, if you fetch OG image and then change the URL, it keeps the old one because the fetch happened earlier. Why would you do a fetch on a url, have its og data prefill, only to change everything after ? It doesn't make sense ?

Users aren't obliged to make sense to developers who are sitting in other places at other times. ;-)  In testing, I'm trying to do things that I could imagine a user doing.  Like pasting the wrong URL and only realising it when the OG displays.

I sketched out a possible workflow response path in #901.

JeffAbrahamson commented:
ac27cfc4ee6 and ec1c49a LGTM, let's at least merge those to reduce PR complexity.

Shriukan33 commented:
Updating this PR : 
* Added a button in creation form to refresh OG data after initial POST, for clumsy users :innocent: 
* Improved responsive on list view, you should be able to see more articles per page on large screen
* Added the description in the list (previously using incorrect attribute)
* Reworded bits of the text you mentioned to be less verbose 
* Increased text limits to 1000 so we're not bothered anymore
* Added og:site_name prefill
* Articles now open in new tabs
* Images are now clickable

Shriukan33 commented:
> [ac27cfc](https://github.com/transport-nantes/tn_web/commit/ac27cfc4ee691824872caa08e9f214f7d4160e8e) and [ec1c49a](https://github.com/transport-nantes/tn_web/commit/ec1c49a2280a629dfcfb0cd9c00ff3117b7fba50) LGTM, let's at least merge those to reduce PR complexity.

Done

Shriukan33 commented:
update :

* Changing URL in creation form now updates the associated opengraph data
* Fixed a minor discrepancy between detailview and listview

Shriukan33 commented on transport_nantes/press/views.py:
```diff
@@ -96,21 +101,99 @@
+        url = request.POST["url"]
+
+    try:
+        response = requests.get(url)
```
This bot is right, it leaves us open to get on any user supplied url.
We would curate the users that have that permission, but still ... I don't see how we could add more security layers.

JeffAbrahamson replied:
That's why such an ajax endpoint should be POST-only and should 403 on GET.  That permits you to use a CSRF token to validate the request.

We don't want to be an OG fetching server, indeed.

JeffAbrahamson replied:
That said, this page is not available except to people with the appropriate permission.  (You did add permissions to pressmention, iirc, right?)  At some point, we are obliged to respond to user input on a website.

JeffAbrahamson replied:
Ah, and it's noting SSRF attacks.  You'll have to google it, but there are ways to validate that it's a valid URL and doesn't include stuff that will lead to further parsing down the line and further action.

Shriukan33 commented:
PR is moved to #910
